src/ebonite/core/objects/metric.py

```diff
@@ -29,3 +36,62 @@ def evaluate(self, truth, prediction):
             return self._function(prediction, truth, **self.args)
         else:
             return self._function(truth, prediction, **self.args)
+
+
+class CallableMetricWrapper:
+    def __init__(self, artifacts: Dict[str, str], requirements: Requirements):
+        self.artifacts = artifacts
+        self.requirements = requirements
+        self.callable = None
+
+    def bind(self, callable):
+        self.callable = callable
+        return self
+
+    @staticmethod
+    def compress(s: bytes) -> str:
+        """
+        Helper method to compress source code
+
+        :param s: source code
+        :return: base64 encoded string of zipped source
+        """
+        zp = zlib.compress(s)
+        b64 = base64.standard_b64encode(zp)
+        return b64.decode('utf8')
+
+    @staticmethod
+    def decompress(s: str) -> bytes:
+        """
+        Helper method to decompress source code
+
+        :param s: compressed source code
+        :return: decompressed source code
+        """
+        zp = base64.standard_b64decode(s.encode('utf8'))
+        src = zlib.decompress(zp)
+        return src
+
+    @classmethod
+    def from_callable(cls, callable):
+        reqs = get_object_requirements(callable)
+        with PickleModelIO().dump(callable) as artifacts:
+            payload = {path: cls.compress(bts) for path, bts in artifacts.bytes_dict().items()}
+        return CallableMetricWrapper(payload, reqs).bind(callable)
+
+    def load(self):
+        with tempfile.TemporaryDirectory() as tmpdir:
+            for path, art in self.artifacts.items():
+                with open(os.path.join(tmpdir, path), 'wb') as f:
+                    f.write(self.decompress(art))
+            self.callable = PickleModelIO().load(tmpdir)
+
+
+class CallableMetric(Metric):
+    def __init__(self, wrapper: CallableMetricWrapper):
+        self.wrapper = wrapper
+
+    def evaluate(self, truth, prediction):
```

Comment:
What about situations when  we have more than 2 metric params? Can it be that instead of truth/prediction, we have something akin of Tuple(truth, optional_param1, optional_param2, etc)/prediction?

Reply:
i thought about it, but for now this can be achieved with custom class and it's fields. and only case i can imagine for different evaluate signature is unsupervised learning where you only have prediction without truth.
maybe in the future we also need to add dataset_type that metric can consume, but for that is another story

Reply:
Well, i just thought of something. What about multiclass classification with macro and micro type of averaging? Or something akin of F-score.